src/player.cpp

```diff
@@ -12567,24 +12564,18 @@ void player::absorb_hit(body_part bp, damage_instance &dam) {
 
         // The worn vector has the innermost item first, so
         // iterate reverse to damage the outermost (last in worn vector) first.
-        for( std::vector<int>::reverse_iterator armor_it = armor_indices.rbegin();
-             armor_it != armor_indices.rend(); ++armor_it ) {
-
-            const int index = *armor_it;
+        for( auto iter = worn.rbegin(); iter != worn.rend(); ) {
+            item& armor = *iter;
 
-            armor_absorb( elem, worn[index] );
+            if( !armor.covers( bp ) ) {
+                ++iter;
+                continue;
+            }
 
-            // now check if armor was completely destroyed and display relevant messages
-            // TODO: use something less janky than the old code for this check
-            if( worn[index].damage >= 5 ) {
-                //~ %s is armor name
-                add_memorial_log(pgettext("memorial_male", "Worn %s was completely destroyed."),
-                                 pgettext("memorial_female", "Worn %s was completely destroyed."),
-                                 worn[index].tname( 1, false ).c_str());
-                add_msg_player_or_npc( m_bad, _("Your %s is completely destroyed!"),
-                                              _("<npcname>'s %s is completely destroyed!"),
-                                              worn[index].tname( 1, false ).c_str() );
-                worn.erase(worn.begin() + index);
+            if( armor_absorb( elem, armor ) ) {
+                iter = decltype(iter)( worn.erase( --iter.base() ) );
```

Comment:
That's actually quite simple, once you understand the reverse iterator class.

The decltype is the typename of the iterator, which is probably something like `std::reverse_iterator<std::list<item>::iterator>` and the `--iter.base()` is because the the reverse iterator points to (i-1), see this helpful answer: http://stackoverflow.com/questions/1830158/how-to-call-erase-with-a-reverse-iterator